Scripts/build/Jenkins/Jenkinsfile

```diff
@@ -0,0 +1,665 @@
+
+    projectsAndUrl.each { projectAndUrl ->
+        if(!fileExists(projectAndUrl.key)) {
+            palMkdir(projectAndUrl.key)
+        }
+        dir(projectAndUrl.key) {           
+            if(fileExists('.git')) {
+                // If the repository after checkout is locked, likely we took a snapshot while git was running,
+                // to leave the repo in a usable state, garbage collect.
+                def indexLockFile = '.git/index.lock'
+                if(fileExists(indexLockFile)) {
+                    palSh('git gc', 'Git GarbageCollect')    
+                }
+                if(fileExists(indexLockFile)) { // if it is still there, remove it
+                    palRm(indexLockFile)
+                }
+            }
+        }
+    }
+
+    def random = new Random()
+    def retryAttempt = 0
+    retry(5) {
+        if (retryAttempt > 0) {
+            sleep random.nextInt(60 * retryAttempt)  // Stagger checkouts to prevent HTTP 429 (Too Many Requests) response from Github
+        }
+        retryAttempt = retryAttempt + 1
+        projectsAndUrl.each { projectAndUrl ->
+            dir(projectAndUrl.key) {
+                def branchName = scm.branches
+                if(projectAndUrl.key == "${ENGINE_REPOSITORY_NAME}") {
+                    branchName = [[name: params.ENGINE_BRANCH]]
+                }
+                checkout scm: [
+                    $class: 'GitSCM',
+                    branches: branchName,
+                    extensions: [
+                        [$class: 'PruneStaleBranch'],
+                        [$class: 'AuthorInChangelog'],
+                        [$class: 'SubmoduleOption', disableSubmodules: disableSubmodules, recursiveSubmodules: true],
+                        [$class: 'CheckoutOption', timeout: 60]
+                    ],
+                    userRemoteConfigs: [projectAndUrl.value]
+                ]
+            }
+        }
+    }
+
+    // CHANGE_ID is used by some scripts to identify uniquely the current change (usually metric jobs)
+    dir(PROJECT_REPOSITORY_NAME) {
+        palSh('git rev-parse HEAD > commitid', 'Getting commit id')
+        env.CHANGE_ID = readFile file: 'commitid'
+        env.CHANGE_ID = env.CHANGE_ID.trim()
+        palRm('commitid')
+            // CHANGE_DATE is used by the installer to provide some ability to sort tagged builds in addition to BRANCH_NAME and CHANGE_ID
+        commitDateFmt = '%%cI'
+        if (env.IS_UNIX) commitDateFmt = '%cI'
+
+        palSh("git show -s --format=${commitDateFmt} ${env.CHANGE_ID} > commitdate", 'Getting commit date')
+        env.CHANGE_DATE = readFile file: 'commitdate'
+        env.CHANGE_DATE = env.CHANGE_DATE.trim()
+        palRm('commitdate')
+    }
+}
+
+def PreBuildCommonSteps(Map pipelineConfig, String repositoryName, String projectName, String pipeline, String branchName, String platform, String buildType, String workspace, boolean mount = true, boolean disableSubmodules = false) {

```

Comment:
nit: Lots of params, no documentation for what values are

Reply:
I didn't find any documentation about these parameters in the existing Jenkins config files (engine and the atom project repo) unfortunately. Might be a request for the infra team to document these parameters since I don't want to provide inaccurate information for them here :(